omc3/scripts/merge_kmod_results.py

```diff
@@ -0,0 +1,237 @@
+import pathlib
+import re
+from functools import reduce
+from operator import mul
+from typing import Dict, List
+from uncertainties import unumpy as up
+
+import tfs
+from generic_parser import EntryPointParameters, entrypoint
+from tfs.tools import significant_digits
+
+from omc3.definitions.constants import PLANES
+from omc3.run_kmod import EXT
+from omc3.run_kmod import LSA_FILE_NAME as LSA_RESULTS
+from omc3.run_kmod import RESULTS_FILE_NAME as RESULTS
+from omc3.utils.logging_tools import get_logger
+
+LOG = get_logger(__name__)
+
+BETASTAR = "BETSTAR"
+ERR = "ERR"
+LABEL = "LABEL"  # Column containing the IP/Beam names
+BEAMS = ("B1", "B2")
+IPS = ("ip1", "ip5")
+LABELS = [f"{ip}{beam}" for ip in IPS for beam in BEAMS]
+
+
+def _validate_for_imbalance(data_frame: tfs.TfsDataFrame) -> bool:
+    """
+    Checks that the provided `TfsDataFrame` contains the expected labels and columns for a
+    luminosity imbalance calculation.
+
+    Args:
+        data_frame (tfs.TfsDataFrame): a loaded `TfsDataFrame` to validate.
+
+    Returns:
+        ``True`` if the provided dataframe is valid, ``False`` otherwise.
+    """
+    expected_labels = set(LABELS)
+    not_found_labels = [
+        label for label in expected_labels if not data_frame[LABEL].str.contains(label).any()
+    ]
+    if any(not_found_labels):
+        return False
+
+    for label in LABELS:
+        if len(data_frame.loc[data_frame[LABEL] == label]) > 1:
+            LOG.error(f"Found label '{label}' several times. Expected only once")
+            raise KeyError(f"Duplicate label '{label}' in dataframe's columns")
+
+    # Validate the columns we need: BETSTAR{X,Y} and ERRBETSTAR{X,Y}
+    expected_columns = [f"{BETASTAR}{p}" for p in PLANES] + [f"{ERR}{BETASTAR}{p}" for p in PLANES]
+    if not all([column in data_frame.columns for column in expected_columns]):
+        return False
+
+    return True
+
+
+def get_lumi_imbalance(data_frame: tfs.TfsDataFrame) -> Dict[str, float]:
+    """
+    Calculate the IP1 / IP5 luminosity imbalance. The luminosity is taken as defined in
+    `Concept of luminosity`, Eq(17): https://cds.cern.ch/record/941318/files/p361.pdf
+
+    The calculation of the luminosity imbalance is then:
+    .. math::
+    \\frac{L_{IP1}}{L_{IP5}}=\\frac{\\sqrt{\\beta_{x1,IP5}+\\beta_{x2,IP5}}\\cdot\\sqrt{\\beta_{y1,IP5}+\\beta_{y2,IP5}}}{\\sqrt{\\beta_{x1,IP1}+\\beta_{x2,IP1}}\\cdot\\sqrt{\\beta_{y1,IP1}+\\beta_{y2,IP1}}}
+
+    Args:
+        data_frame (tfs.TfsDataFrame): a `TfsDataFrame` with the results from a kmod analysis.
+
+    Returns:
+        A dictionary with the imbalance, the betas at IPs and their errors.
+    """
+
+    lumi_coefficient: Dict[str, float] = {}
+
+    for ip in IPS:
+        LOG.debug(
+            f"Computing lumi contribution from optics for IP {ip}"
+        )
+        ip_row = data_frame.loc[data_frame[LABEL].str.startswith(ip)]
```

Comment:
it should be `ip_rows` as it is two (one for each beam). Otherwise the `.sum()` below is very confusing.
Maybe even `rows_beams_in_ip`